### PR TITLE
Label selectors for the `AuthConfig`

### LINKED ISSUE #174: Label selectors for `AuthConfig`s
Similarly to `Secret` label selectors that allow to [filter](https://github.com/Kuadrant/authorino/blob/cc0a0b4ae232550027931615a4e55277051e4c44/controllers/secret_controller.go#L116) which Kubernetes `Secret` resources are watched by an Authorino instance (to be reconciled as possible Authorino API keys tied to an `AuthConfig`), it should be able to filter the `AuthConfig`s themselves as well.

By adding support for this same type of resource filtering based on labels, it will be possible to split sets of `AuthConfig`s to be watched by different instances of Authorino monitoring a same space, either multiple cluster-wide instances watching the entire cluster or multiple namespaced instances in a given namespace.

An obvious advantage is to enable sharding of `AuthConfig`s while preventing unnecessary resources from being cached by instances configured to never receive traffic related to those corresponding hosts.

This can be a way into supporting more fine-grained multi-tenancy, beyond the separation by Kubernetes namespaces.

A side effect (not necessarily a positive one) is that it will open up for having `AuthConfig`s defined in the cluster, eventually even reconciled in the past by some Authorino instance (i.e. `authconfig/status.Ready=True`), that end up becoming stale due to a change on the label selectors of the instance. (This is analogous to having an `AuthConfig` created in a namespace where once a namespaced instance of Authorino was running and then got deleted or moved to another namespace, but elevated to cluster-wide instances as well.)

To ensure concurrent instances (_Attention!_ I don't mean concurrent _replicas_ of the same instance) watching different sets of `AuthConfig`s don't end up sharing the same leader, it is imperative that Authorino uses the value of the `AuthConfig` label selectors to generating the `LeaderElectionID`, instead of having it hardcoded in: https://github.com/Kuadrant/authorino/blob/cc0a0b4ae232550027931615a4e55277051e4c44/main.go#L147

The value of the label selectors can be used as the source of a hash to be interpolated within `LeaderElectionID`. Instances and replics eventually watching the same set of `AuthConfig`s will share all the same (and only) leader, whereas independent/isolated instances (i.e. watching different sets of `AuthConfig`s) will have also independent leaders (to take care of the status update).

----

Mentioned in https://github.com/Kuadrant/authorino/issues/129#issuecomment-947844871.

### PR DESCRIPTION
**Pull request by guicassolato**
Filters `AuthConfig`-related events by labels of the resource.

Labels selectors for the instance can be set via newly introduced `AUTH_CONFIG_LABEL_SELECTOR` environment variable, in the format of ~zero or more `expected-label=expected-value` key-value pairs, separated by anything that qualifies as a space in Unicode (i.e. `\t`, `\n`, `\v`, `\f`, `\r`, ` `, `0x85`, `0xA0`)~ parseable [Kubernetes label selectors](https://pkg.go.dev/k8s.io/apimachinery/pkg/labels#Parse).

All the following are valid examples of `AuthConfig` label selector filters:

```
AUTH_CONFIG_LABEL_SELECTOR="authorino.3scale.net/managed-by=authorino"
AUTH_CONFIG_LABEL_SELECTOR="authorino.3scale.net/managed-by=authorino,other-label=other-value"
AUTH_CONFIG_LABEL_SELECTOR="authorino.3scale.net/managed-by in (authorino,kuadrant)"
AUTH_CONFIG_LABEL_SELECTOR="authorino.3scale.net/managed-by!=authorino-v0.4"
AUTH_CONFIG_LABEL_SELECTOR="!disabled"
```

By default, no label is required in the `AuthConfig`s for them to be watched by the Authorino controller – i.e. `AUTH_CONFIG_LABEL_SELECTOR` is empty and therefore the controller will reconcile all `AuthConfig`s in the watched space.

**Breaking changes:**
The environment variable `AUTHORINO_SECRET_LABEL_KEY`, used to specify labels required in `Secret`s to be watched by Authorino, was renamed to `SECRET_LABEL_SELECTOR` and now also expects the same format of zero or more key-value pairs as `AUTH_CONFIG_LABEL_SELECTOR`.

The name of `ConfigMap` created to handle the leader election is now determined by the value of `AUTH_CONFIG_LABEL_SELECTOR`.

**Bug fixes:**
The PR also fixes a bug in the label filter of the `Secret` controller that caused removed required labels from existing `Secret`s not to trigger the reconciliation of the `Secret` and related `AuthConfig`.

The PR also fixes a bug in the selection of `AuthConfig`s related to `Secret`s reconciliated, to trigger the reconciliation of the `AuthConfig`s. Now the sets of labels expected in the `Secret` and matching `spec.identity.apiKey.labelSelectors` do not have to be identical (checked before with `reflect.DeepEqual`), but to match as in regular label selector matches performed by Kubernetes, like when the `Secret`s are listed and added to cache of valid credentials of an identity config: https://github.com/Kuadrant/authorino/blob/6e741f722d4acb10c461b9c430dddf7c28039c35/pkg/config/identity/api_key.go#L56

As as consequence of the above, the default label selectors used to filter `Secret`-related events to the reconciler no longer has to be mentioned in `spec.identity.apiKey.labelSelectors`.

**Logging:**
Adds a new debug log entry to print instance configuration options during bootstrap, and fixes other log entries related to the reconciliation of `AuthConfig`s, `Secret`s and `AuthConfig` status updates, especially when skipped due to non-watched resource.

**Doc updates:**
Description of the API key feature in the examples and _Architecture_ page updated to reflect the default label selectors used to filter `Secret`-related events to the reconciler no longer needed to be mentioned in `spec.identity.apiKey.labelSelectors`.

----

Closes #174

### CODE REVIEW

File: controllers/label_selector.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	for key, value := range selector {
+		if _, ok := labels[key]; !ok || labels[key] != value {
+			return false
+		}
+	}
+
+	return true
+}
+
+func FilterByLabels(selector map[string]string) predicate.Funcs {
```
[the remark this answers is not in the document]
Comment (guicassolato) [reply controllers/label_selector.go]: Done, @eguzki. Thanks for the tip!

File: controllers/auth_config_controller.go [hunk 2]
```diff
@@ -459,6 +455,7 @@ func (r *AuthConfigReconciler) translateAuthConfig(ctx context.Context, authConf
 func (r *AuthConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&configv1beta1.AuthConfig{}).
+		WithEventFilter(FilterByLabels(r.LabelSelector)).
```
[the remark this answers is not in the document]
Comment (guicassolato) [reply controllers/auth_config_controller.go]: Done

**Comment by eguzki** [comment]
So, let me explain what I understand to check if I understood correctly. 

Atuhorino will have a deployment time env var called `SECRET_LABEL_SELECTOR` and all secrets matching the label selector will be monitored and added to the cache. Eventually, this `SECRET_LABEL_SELECTOR` will be configurable in the spec of the authorino CRD from the operator. 

When I (as a authorino client) create an `AuthConfig` for my API KEY based authN, I have basically two options.

A) create the secret with labels matching `SECRET_LABEL_SELECTOR`  selector, then the authconfig looks like this
```yaml
spec:
  identity:
  - apiKey: {}
    credentials:
      in: authorization_header
      keySelector: APIKEY
    name: MyAPIKey
```
Note that I do not specify the label selector in the `apiKey` object.

B) create the secret with my own labels, i.e. `MYAPP=MYLABEL`, then create authconfig like this:
```yaml
spec:
  identity:
  - apiKey:
      labelSelectors:
        MYAPP: MYLABEL
    credentials:
      in: authorization_header
      keySelector: APIKEY
    name: MyAPIKey
```

With option A, any change done in the secret will be reconciled and authorino will enforce new api keys (provided the value in the secret changed).
With option B, the secret will not be reconciled. So authorino will read initially the secret, but changes will not be reconciled. So, the recommended option would be A). 

Am I right?

**Comment by eguzki** [comment]
The code LGTM

**Comment by guicassolato** [comment]
> Am I right?

Almost 100%, but not quite 🙂 Let me try to explain...

> Atuhorino will have a deployment time env var called `SECRET_LABEL_SELECTOR` and all secrets matching the label selector will be monitored

Correct.

>[...] and added to the cache.

Incorrect.

> Eventually, this `SECRET_LABEL_SELECTOR` will be configurable in the spec of the authorino CRD from the operator.
> 

Correct. It is right here: https://github.com/Kuadrant/authorino-operator/blob/6bd2a13104ecb99841dbdf9f040717d29d669fd6/api/v1beta1/authorino_types.go#L78

----

The `SECRET_LABEL_SELECTOR` env var already exists, except that right now it is called `AUTHORINO_SECRET_LABEL_KEY` and it only expects a label key and no label value. This PR changes its name and format, but it doesn't change its meaning.

`SECRET_LABEL_SELECTOR` is (and will continue to be) used to filter the events related to `Secret`s that reach the reconciler. Period.

As before, we should use `SECRET_LABEL_SELECTOR` to avoid the reconciler being overwhelmed with lots of events that are completely irrelevant to the `AuthConfig`s being managed. (Imagine a cluster-wide deployment of Authorino watching for every single `Secret`-related event across the entire cluster. This would be neither efficient, nor safe.)

---

_A priori_, `spec.identity.apiKey.labelSelectors` has nothing to do with `SECRET_LABEL_SELECTOR`. To understand that, we need to understand what each of Authorino's reconcilers do regarding the reconciliation and caching of API keys:
- `AuthConfigReconciler` selects all the `Secret`s that are supposed to be represent API keys. `identity.APIKey.GetCredentialsFromCluster(...)` lists all `Secret`s whose labels match `AuthConfig.spec.identity.apiKey.labelSelectors` and caches them as API keys.

- `SecretReconciler` goes to through the entire cache of `AuthConfig`s and, for every `AuthConfig` whose `spec.identity.apiKey.labelSelectors` matches the labels of the `Secret` being reconciled, it triggers the reconciliation of the `AuthConfig`, calling `AuthConfigReconciler.Reconcile(...)`. It is then `AuthConfigReconciler` that will reconcile the `AuthConfig` and refresh the cache of API keys; `SecretReconciler` does not write to the cache.

<br/>

**What happens if I have `SECRET_LABEL_SELECTOR=x` and an `AuthConfig` where `spec.identity.apiKey.labelSelectors=y`?**

- Whenever the `AuthConfig` is reconciled, all `Secret`s where `y ⊆ metadata.labels` will be cached.
- Only events related to `Secret`s where `x ⊆ metadata.labels` will be watched by the `SecretReconciler`; among those, only when `y ⊆ metadata.labels` the `AuthConfig` will be reconciled and therefore the event reflected in the cache of API keys of the `AuthConfig`
    - If the `x ⊈ metadata.labels` but `y ⊆ metadata.labels` for a given `Secret`, then events related to this `Secret` will not be watched and the potentially impacted cache of `AuthConfig` will only be updated by other means (e.g. if something touches the `AuthConfig` itself, rather than the `Secret`)
    - If the `x ⊆ metadata.labels` but `y ⊈ metadata.labels` for a given `Secret`, then `SecretReconciler` will watch the event, but it won't identify the `AuthConfig` whose `spec.identity.apiKey.labelSelectors=y` as potentially impacted and therefore it won't trigger the reconciliation of that `AuthConfig`, causing its cache of API keys to remain as-is
- If you want to ensure an `AuthConfig` to cache only `Secret`s whose events will be watched by `SecretReconciler` and yet give it, in the labels of its related `Secret`s, something that allows to distinguish this `AuthConfig` from others, set `spec.identity.apiKey.labelSelectors=x ∪ y` and, of course, make sure `x ∪ y ⊆ metadata.labels` for every related `Secret`.
   - If you have an instance of Authorino forever meant to watch only one `AuthConfig`, set `SECRET_LABEL_SELECTOR=spec.identity.apiKey.labelSelectors=x` and, of course, `x ⊆ metadata.labels` should be true for every related `Secret`.

<br/>

> When I (as a authorino client) create an `AuthConfig` for my API KEY based authN, I have basically two options.
> 
> A) create the secret with labels matching `SECRET_LABEL_SELECTOR` selector, then the authconfig looks like this
> 
> ```yaml
> spec:
>   identity:
>   - apiKey: {}
>     credentials:
>       in: authorization_header
>       keySelector: APIKEY
>     name: MyAPIKey
> ```
> 
> Note that I do not specify the label selector in the `apiKey` object.
>
> [...] any change done in the secret will be reconciled and authorino will enforce new api keys (provided the value in the secret changed).

Incorrect.

- In this case, all `Secret`s in the entire space watched by Authorino are potential API keys valid to authenticate to this service. `AuthConfigReconciler` will make sure to cache them all.
- However, only events related to the `Secret`s where `SECRET_LABEL_SELECTOR ⊆ metadata.labels` will trigger the reconciliation of the `AuthConfig`
   - If you have a `Secret` where `SECRET_LABEL_SELECTOR ⊈ metadata.labels` and that `Secret` is updated, the update will not be immediately reflected in the cache of the `AuthConfig`; only when some other `Secret` where `SECRET_LABEL_SELECTOR ⊆ metadata.labels` is updated the cache will be refreshed (entirely)

> B) create the secret with my own labels, i.e. `MYAPP=MYLABEL`, then create authconfig like this:
> 
> ```yaml
> spec:
>   identity:
>   - apiKey:
>       labelSelectors:
>         MYAPP: MYLABEL
>     credentials:
>       in: authorization_header
>       keySelector: APIKEY
>     name: MyAPIKey
> ```
> 
> [...] the secret will not be reconciled. So authorino will read initially the secret, but changes will not be reconciled.

Correct.

> So, the recommended option would be A).

The recommended option is always a mix of option A and option B. As long as you understand what `SECRET_LABEL_SELECTOR` does (limits the scope within the reconciliation space) and what `AuthConfig.spec.identity.apiKey.labelSelectors` does (selects `Secret`s as valid API keys whenever the `AuthConfig` is reconciled), you're good.

To be safe, I'd look for the following configuration:
```
SECRET_LABEL_SELECTOR ⊆ AuthConfig.spec.identity.apiKey.labelSelectors ⊆ Secret.metadata.labels
```

To be even safer and have more flexibility with the sets of `AuthConfig`s and `Secret`s:
```
SECRET_LABEL_SELECTOR ⊂ AuthConfig.spec.identity.apiKey.labelSelectors ⊂ Secret.metadata.labels
```

**Comment by eguzki** [comment]
Super clear, thanks. 

I was thinking then that this statement should be in documentation (maybe it is and I missed it):
```
SECRET_LABEL_SELECTOR ⊂ AuthConfig.spec.identity.apiKey.labelSelectors ⊂ Secret.metadata.labels
```

I will make sure that kuadrant-controller works accordingly.

**Review by eguzki — APPROVED** [review]

**Pull request merged**